From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrzej "retrofaza" Subocz <andrzej@subocz.pl>
Date: Wed, 19 Aug 2026 15:24:14 +0200
Subject: [PATCH 46/46] NListtree.mcc: bound tree image strings to the destination

Backport upstream AROS commit 6897aa2dc5d5ad15bae201c0ec1d26ffbb0f5b79.

strncat() with DATA_BUF_SIZE as its third argument limits how much is read
from the source; it does not account for bytes already present in data->buf.
When the tree image string was assembled near the end of that buffer it
overwrote the header of the adjacent TLSF block. Bellatrix sees precisely
that damage during the Wanderer boot: REMOVE_HEADER receives a NULL block
from a corrupted free list shortly after NListtree is loaded.

Restore the upstream NList implementation: format each individual image
specifier into a small temporary buffer and append it with strlcat(), whose
bound describes the destination. Besides fixing ISSUE-0037, this removes an
unrelated heap failure from vc4gfx QEMU and hardware validation.
---
 .../zune/nlist/nlisttree_mcc/NListtree.c      | 28 +++++++++----------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/workbench/classes/zune/nlist/nlisttree_mcc/NListtree.c b/workbench/classes/zune/nlist/nlisttree_mcc/NListtree.c
index 9eccf4d02f..c8eb22a8e0 100644
--- a/workbench/classes/zune/nlist/nlisttree_mcc/NListtree.c
+++ b/workbench/classes/zune/nlist/nlisttree_mcc/NListtree.c
@@ -2795,15 +2795,15 @@ static void InsertTreeImages( struct NListtree_Data *data, struct MUI_NListtree_
 
     if ( ( x1 != -1 ) && x2 )
     {
-      char tmpbuf[DATA_BUF_SIZE];
+      char tmp[128];
 
       if ( x1 == SPEC_Space )
         otn->tn_Space += x2 + data->IndentWidth;
 
       if ( otn->tn_Space > 0 )
       {
-        snprintf(tmpbuf, DATA_BUF_SIZE, "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)SPEC_Space, (int)otn->tn_Space );
-        strncat(data->buf, tmpbuf, DATA_BUF_SIZE);
+        snprintf(tmp, sizeof(tmp), "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)SPEC_Space, (int)otn->tn_Space);
+        strlcat(data->buf, tmp, DATA_BUF_SIZE);
 
         otn->tn_ImagePos += otn->tn_Space;
         otn->tn_Space = 0;
@@ -2818,8 +2818,8 @@ static void InsertTreeImages( struct NListtree_Data *data, struct MUI_NListtree_
         */
         if(isFlagSet(data->Flags, NLTF_NO_ROOT_TREE) && gp->tn_Parent == NULL)
         {
-          snprintf(tmpbuf, DATA_BUF_SIZE, "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)SPEC_Space, (int)x2 );
-          strncat(data->buf, tmpbuf, DATA_BUF_SIZE);
+          snprintf(tmp, sizeof(tmp), "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)SPEC_Space, (int)x2);
+          strlcat(data->buf, tmp, DATA_BUF_SIZE);
 
           otn->tn_ImagePos += x2;
         }
@@ -2830,8 +2830,8 @@ static void InsertTreeImages( struct NListtree_Data *data, struct MUI_NListtree_
           if(data->IndentWidth > 0)
             x2 += 2;
 
-          snprintf(tmpbuf, DATA_BUF_SIZE, "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)x1, (int)x2 );
-          strncat(data->buf, tmpbuf, DATA_BUF_SIZE);
+          snprintf(tmp, sizeof(tmp), "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)x1, (int)x2);
+          strlcat(data->buf, tmp, DATA_BUF_SIZE);
 
           otn->tn_ImagePos += x2;
         }
@@ -2843,7 +2843,7 @@ static void InsertTreeImages( struct NListtree_Data *data, struct MUI_NListtree_
 static void InsertImage( struct NListtree_Data *data, struct MUI_NListtree_TreeNode *otn )
 {
   struct MUI_NListtree_ListNode *ln = CLN( otn );
-  char tmpbuf[DATA_BUF_SIZE];
+  char tmp[128];
   LONG x1 = -1;
 
   InsertTreeImages( data, otn, otn, 0 );
@@ -2859,22 +2859,22 @@ static void InsertImage( struct NListtree_Data *data, struct MUI_NListtree_TreeN
       x1 = IMAGE_Closed;
     }
 
-    snprintf(tmpbuf, DATA_BUF_SIZE, "\033O[%p]", (APTR)data->Image[x1].ListImage );
-    strncat(data->buf, tmpbuf, DATA_BUF_SIZE);
+    snprintf(tmp, sizeof(tmp), "\033O[%p]", (APTR)data->Image[x1].ListImage);
+    strlcat(data->buf, tmp, DATA_BUF_SIZE);
 
     x1 = SPEC_Hor;
 
     // add some indent width
     if(data->IndentWidth > 0)
     {
-      snprintf(tmpbuf, DATA_BUF_SIZE, "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)x1, (unsigned int)data->IndentWidth);
-      strncat(data->buf, tmpbuf, DATA_BUF_SIZE);
+      snprintf(tmp, sizeof(tmp), "\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)x1, (int)data->IndentWidth);
+      strlcat(data->buf, tmp, DATA_BUF_SIZE);
     }
 
     if(data->UseFolderImage == TRUE)
     {
-      snprintf(tmpbuf, DATA_BUF_SIZE, "\033O[%p]\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Folder].ListImage, (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (unsigned int)SPEC_Space, 3);
-      strncat(data->buf, tmpbuf, DATA_BUF_SIZE);
+      snprintf(tmp, sizeof(tmp), "\033O[%p]\033O[%p;%x;%d,%d]", (APTR)data->Image[IMAGE_Folder].ListImage, (APTR)data->Image[IMAGE_Tree].ListImage, (unsigned int)MUIA_TI_Spec, (int)SPEC_Space, (int)3);
+      strlcat(data->buf, tmp, DATA_BUF_SIZE);
     }
   }
 }
-- 
2.43.0
